onnx/defs/math/defs.cc

```diff
@@ -3337,4 +3337,220 @@ ONNX_OPERATOR_SET_SCHEMA(
+  const auto U_idx = 0;
+  const auto S_idx = compute_uv ? 1 : 0;
+  const auto Vh_idx = 2;
+
+  // All output elements types are same as input element types
+  if (compute_uv) {
+    propagateElemTypeFromInputToOutput(ctx, 0, U_idx);
+    propagateElemTypeFromInputToOutput(ctx, 0, Vh_idx);
+  }
+
+  propagateElemTypeFromInputToOutput(ctx, 0, S_idx);
+
+  // SVD is applied to lowest two dimensions of the matrix.
+  // The rest of the dimensions are used for broadcasting,
+  // see https://numpy.org/doc/stable/reference/generated/numpy.linalg.svd.html
+
+  const auto A_shape = ctx.getInputType(0)->tensor_type().shape();
+  const auto A_dim_size = A_shape.dim_size();
+
+  if (A_dim_size < 2) {
+    fail_shape_inference("Input tensor must have at least 2 dimensions");
+  }
+
+  auto dim_idx = 0;
+
+  auto U_shape = compute_uv ? ctx.getOutputType(U_idx)->mutable_tensor_type()->mutable_shape() : nullptr;
+  auto S_shape = ctx.getOutputType(S_idx)->mutable_tensor_type()->mutable_shape();
+  auto Vh_shape = compute_uv ? ctx.getOutputType(Vh_idx)->mutable_tensor_type()->mutable_shape() : nullptr;
+
+  // Copy over all dimensions but the last two
+  for (; dim_idx < A_dim_size - 2; ++dim_idx) {
+    const auto dim = A_shape.dim(dim_idx);
+
+    if (compute_uv) {
+      *U_shape->add_dim() = dim;
+      *Vh_shape->add_dim() = dim;
+    }
+
+    *S_shape->add_dim() = dim;
+  }
+
+  // SVD Factorization is only applied to the last two dimensions.
+
+  const auto A_n_rows_dim = A_shape.dim(dim_idx);
+  const auto A_n_cols_dim = A_shape.dim(dim_idx + 1);
+
+  const auto min_dimension = A_n_rows_dim.dim_value() < A_n_cols_dim.dim_value() ? A_n_rows_dim : A_n_cols_dim;
+
+  if (compute_uv) {
+    const auto* full_matrices_attr = ctx.getAttribute("full_matrices");
+    const auto full_matrices = full_matrices_attr == nullptr || full_matrices_attr->i() != 0;
+
+    if (full_matrices) {
+      // Full SVD
+      // U and Vh are both square matrices
+      *U_shape->add_dim() = A_n_rows_dim;
+      *U_shape->add_dim() = A_n_rows_dim;
+
+      *S_shape->add_dim() = min_dimension;
+
+      *Vh_shape->add_dim() = A_n_cols_dim;
+      *Vh_shape->add_dim() = A_n_cols_dim;
+    } else {
+      // Partial SVD
+      // U's number of columns becomes the min dimension
+      // Vh's number of rows becomes the min dimension
+      *U_shape->add_dim() = min_dimension;
+      *U_shape->add_dim() = A_n_rows_dim;
+
+      *S_shape->add_dim() = min_dimension;
+
+      *Vh_shape->add_dim() = min_dimension;
+      *Vh_shape->add_dim() = A_n_cols_dim;
+    }
+  } else {
+    *S_shape->add_dim() = min_dimension;
+  }
+}
+
+ONNX_OPERATOR_SET_SCHEMA(
+    SVD,
+    18,
+    OpSchema()
+        .SetDoc(SVD_doc)
+        .Input(
+            0,
+            "A",
+            "Input tensor to factor. Must have at least 2 dimensions.",
+            "T",
+            OpSchema::Single,
+            true,
+            1,
+            // TODO
+            OpSchema::Unknown)
```

Comment:
x-posting from LFAI slack

 My basic understanding is that SVD (and other matrix factorizations) are usually differentiable but sometimes are numerically unstable. I'm not sure exactly how this maps to specifying differentiability in the onnx operator inputs/outputs -- should all input and output differentiability be unknown?
From the pytorch docs re: SVD differentiability
Gradients computed using U or Vh will only be finite when A does not have repeated singular values. If A is rectangular, additionally, zero must also not be one of its singular values. Furthermore, if the distance between any two singular values is close to zero, the gradient will be numerically unstable ...
torch.linalg.svdvals() computes only the singular values. Unlike torch.linalg.svd(), the gradients of svdvals() are always numerically stable.
https://pytorch.org/docs/stable/generated/torch.linalg.svd.html#torch.linalg.svd
https://pytorch.org/docs/stable/generated/torch.linalg.svdvals.html
https://pytorch.org/docs/stable/generated/torch.linalg.eigh.html
Previous discussion in pytorch:
https://github.com/pytorch/pytorch/issues/49886
https://github.com/pytorch/pytorch/issues/57272
https://github.com/pytorch/pytorch/issues/47599#issuecomment-812539850

Reply:
I'd say let's make it differentiable following Pytorch.
```python
import torch
A = torch.ones(3, 3).requires_grad_()
U, S, V = torch.linalg.svd(A, full_matrices=False)
print(U.requires_grad) # True, so differentiable.
print(S.requires_grad) # True, so differentiable.
print(V.requires_grad) # True, so differentiable.
(U+S+V).relu().sum().backward()
print(A.grad) # A is differentiable because A.grad is
                     # tensor([[ 1.4211e+00,  1.9990e+07, -1.9990e+07],
                     #         [-1.9990e+07, -5.9821e-01,  1.9990e+07],
                     #         [ 1.9990e+07, -1.9990e+07,  7.3364e-01]])
```

Reply:
so to clarify, the issues with numerical stability and other sidecases with computing the gradient should not impact if the inputs and outputs are marked as differentiable in ONNX? 

i.e. example of 0 singular value for rectangular A

```python
A = torch.ones(4, 3).requires_grad_()
U, S, V = torch.linalg.svd(A, full_matrices=False)
print(S)
print(U.requires_grad)
print(S.requires_grad)
print(V.requires_grad)
(U + torch.cat((S + V, torch.zeros(1, 3)))).relu().sum().backward()
print(A.grad)

# tensor([3.4641e+00, 1.6859e-07, 0.0000e+00], grad_fn=<LinalgSvdBackward0>)
# True
# True
# True
# tensor([[nan, nan, nan],
        [nan, nan, nan],
        [nan, nan, nan],
        [nan, nan, nan]])
```

Sorry to be pedantic but I'm not entirely clear from the onnx docs if this passes the criteria 

```
Differentiable inputs, which can be referenced in Gradient's xs attribute.
Differentiable outputs, which can be referenced in Gradient's y attribute.
```

Reply:
Re: how to compute SVD's backward, what level of detail are you looking for in the PR? SVD's backward is pretty gnarly and I'm probably not qualified to go too far in depth on it. pytorch has some docs on it here https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/FunctionsManual.cpp#L3236-L3348 and uses https://j-towns.github.io/papers/svd-derivative.pdf jax's docs reference https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf for the full matrix case

Reply:
Let's consider a simple function `relu`. It's not differentiable at `0` point, right? We just return a reasonable sub-differentiable value for `0`. We don't mark `relu` non-differentiable even `0` happens quite often. Similarly, marking `svd` as differentiable is reasonable to me. If a function is differentiable at most areas, I'd mark it as differentiable.

`Gradient` op in ONNX computes `[dy/dx0, dy/dx1, ..., dy/dxn]`. When Pytorch call `.backward()`, it computes `dout/din` for each op, so `svd`'s output can be referenced by `y` attribute and `svd`'s inputs can be referenced by `xs` attribute.

Reply:
Ah ok thank you so much, makes a ton of sense

Reply:
Actually I'm going to do a bit more reading on how the backward pass is calculated so I can add documentation about the numerical stability to the op doc. Will see if I pick up enough to add to PR

Reply:
@wschin I put together some documentation and example implementations of SVDs derivative in this notebook https://github.com/williamberman/svd-derivative/blob/main/svd-derivative.ipynb . Will add to PR description